Source/Core/VideoCommon/VertexShaderManager.cpp

```diff
@@ -407,10 +407,14 @@ void VertexShaderManager::SetConstants(const std::vector<std::string>& textures)
     if (g_freelook_camera.IsActive() && xfmem.projection.type == ProjectionType::Perspective)
       corrected_matrix *= g_freelook_camera.GetView();
 
+    GraphicsModActionData::Input::Projection input{std::move(corrected_matrix)};
+    GraphicsModActionData::Output::Projection output;
     for (auto action : projection_actions)
     {
-      action->OnProjection(&corrected_matrix);
+      action->OnProjection(input, &output);
+      input.matrix = std::move(output.matrix);
```

Comment:
This input/output split here seems not entirely thought through. Every `OnProjection()` implementation now needs to write into `output->matrix` even if it doesn't modify the matrix at all, which is not only a lot of needless copying, it's also easy to forget. For example...